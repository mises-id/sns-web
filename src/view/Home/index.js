/*
 * @Author: lmk
 * @Date: 2021-07-10 16:12:04
 * @LastEditTime: 2022-03-07 14:53:31
 * @LastEditors: lmk
 * @Description:
 */
import React, { useEffect, useState } from "react";
import { useTranslation } from "react-i18next";
import { Badge, Tabs } from "zarm";
import send from "@/images/send.png";
import "./index.scss";
import Image from "@/components/Image";
import { getUserSelfInfo } from "@/api/user";
import { useDispatch, useSelector } from "react-redux";
import { setLoginForm, setUserAuth } from "@/actions/user";
import { urlToJson } from "@/utils";
import { useDidRecover } from "react-router-cache-route";
const { Panel } = Tabs;
const Home = ({ history, children = [] }) => {
  const { t } = useTranslation();
  const [tab, setTab] = useState([
    { path: "/home/following", text: t("follow"), badge: 0 },
    { path: "/home/discover", text: t("discover") },
    { path: "/home/me", text: t("me") },
  ]);
  const [value, setvalue] = useState(0);
  const dispatch = useDispatch();
  const setTabActive = () => {
    const { pathname } = window.location;
    //get hostname to set tabIndex
    // const path = pathname;
    switch (pathname) {
      case "/home/following":
        setvalue(0);
        break;
      case "/home/discover":
        setvalue(1);
        break;
      default:
        getUserInfo();
        setvalue(2);
        break;
    }
    const { mises_id, nonce, sig } = urlToJson();
    if (mises_id && nonce && sig) {
      const auth = `mises_id=${mises_id}&nonce=${nonce}&sig=${sig}`;
      dispatch(setUserAuth(auth));
    }
  };
  const user = useSelector((state) => state.user) || {};
  const { auth, token, badge = {},isFistLogin,loginForm } = user;
  // useEffect(() => {
  //   auth &&
  //     !token &&
  //     signin({
  //       provider: "mises",
  //       user_authz: { auth },
  //     }).then((data) => {
  //       data.token && dispatch(setUserToken(data.token));
  //       // localStorage.setItem('setAccount',true)
  //     });
    
  //   console.log(auth, "--------------------", token);
  // }, [auth, token]); // eslint-disable-line react-hooks/exhaustive-deps
  useEffect(() => {
    if(isFistLogin&&!loginForm.is_airdropped){
      history.push('/airdrop')
    }
    // eslint-disable-next-line
  }, [isFistLogin,loginForm])
  
  const getUserInfo = () => {
    if (auth && token) {
      return getUserSelfInfo().then((res) => {
        dispatch(setLoginForm(res));
      });
    }
    return Promise.resolve()
  };
  useDidRecover(() => {
    window.refreshByCacheKey("/follow");
    window.refreshByCacheKey("/notifications");
    window.refreshByCacheKey("/myLikes");
    window.refreshByCacheKey("/myPosts");
    window.refreshByCacheKey("/blackList");
    window.refreshByCacheKey("/userFollowPage");
    window.refreshByCacheKey("/post");
    window.refreshByCacheKey("/comment");
    window.refreshByCacheKey("/userInfo");
  });
  useEffect(() => {
    document.body.style.overflow = "hidden";
    setTabActive();
    const listen = history.listen(setTabActive);
    return () => {
      document.body.style.overflow = "auto";
      listen();
    };
  }, []); // eslint-disable-line react-hooks/exhaustive-deps
  useEffect(() => {
    tab[0].badge = badge.total;
    setTab([...tab]);
  }, [badge.total]); // eslint-disable-line react-hooks/exhaustive-deps
  //change current and change route
  const getChange = (val) => {
    setvalue(val);
    const path = tab[val];
    history.push(path.path);
  };
  const createPosts = () => {
    history.push({ pathname: "/createPosts" });
  };
  //Show current route
  const showChild = (path) =>
    children.find((val) => val.key === path) || <div></div>;
  return (
    <div>
      <Tabs value={value} onChange={getChange} lineWidth={10} swipeable={false}>
        {tab.map((val, index) => (
          <Panel
            key={val.path}
            title={
              <>
                <span className={value === index ? "active" : "unactive"}>
                  {val.text}
                </span>
                {val.badge > 0 && (
                  <Badge shape="circle" className="badge" text={val.badge} />
                )}
              </>
            }
          >
            {showChild(val.path)}
          </Panel>
        ))}
      </Tabs>
      {token && (
        <div className="m-position-fixed createPosts">
          <Image
            size={75}
            alt="posts"
            onClick={createPosts}
            source={send}
            shape="square"
          ></Image>
        </div>
      )}
    </div>
  );
};
export default Home;
