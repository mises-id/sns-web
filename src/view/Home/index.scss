.active{
  font-size: 19px;
  font-weight: bolder;
  transition: all .2s;
  color: #333333;
}
.unactive{
  color: #666666;
  font-size: 15px;
}
.za-tabs--horizontal .za-tabs__tablist{
  height: 50px;
  .za-tabs__tab{
    position: relative;
    .badge{
      position: absolute;
      top: 6px;
      right: 7px;
    }
  }
}
.za-tabs--horizontal .za-tabs__line{
  height: 4px;
  margin-bottom: 2px;
  .za-tabs__line__inner{
    border-radius: 10px;
  }
}
.za-tabs{
  .za-tabs__header{
    background-color: white;
    &::after{
      display: none;
    }
  }
}
.createPosts{
  right: 15px;
  z-index: 99;
  bottom: 30px;
}
.Panel{
  padding-top: 10px;
}
// 生成一个背景色为#5D61FF,宽高为55px,下边距为10px的圆形
.airdrop{
  margin: 0 auto;
  color: white;
  line-height: 55px;
  text-align: center;
  font-size: 12px;
  width: 55px;
  height: 55px;
  background: linear-gradient(45deg, #5D61FF, #5D78FF);
  box-shadow: 0px 5px 10px 0px rgba(93, 97, 255, 0.3);
  border-radius: 50%;
  font-weight: 600;
}