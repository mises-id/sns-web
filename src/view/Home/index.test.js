import React from 'react';
import Home from './';
import {shallow } from 'enzyme';
describe('Home', () => {
  it('should render correctly', () => {
    const component = shallow(<Home></Home>)
    expect(component).toMatchSnapshot();
  });
});