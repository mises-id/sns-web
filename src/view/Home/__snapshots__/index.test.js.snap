// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should render correctly 1`] = `
<Fragment>
  <Tabs
    direction="horizontal"
    disabled={false}
    lineWidth={10}
    onChange={[Function]}
    prefixCls="za-tabs"
    scrollable={false}
    swipeable={false}
    value={0}
  >
    <TabPanel
      key="/home/following"
      prefixCls="za-tabs"
      title={
        <React.Fragment>
          <span
            className="active"
          >
            follow
          </span>
        </React.Fragment>
      }
    >
      <div />
    </TabPanel>
    <TabPanel
      key="/home/discover"
      prefixCls="za-tabs"
      title={
        <React.Fragment>
          <span
            className="unactive"
          >
            discover
          </span>
        </React.Fragment>
      }
    >
      <div />
    </TabPanel>
    <TabPanel
      key="/home/recent"
      prefixCls="za-tabs"
      title={
        <React.Fragment>
          <span
            className="unactive"
          >
            recent
          </span>
        </React.Fragment>
      }
    >
      <div />
    </TabPanel>
    <TabPanel
      key="/home/me"
      prefixCls="za-tabs"
      title={
        <React.Fragment>
          <span
            className="unactive"
          >
            me
          </span>
        </React.Fragment>
      }
    >
      <div />
    </TabPanel>
  </Tabs>
</Fragment>
`;
